testhelpers/docker_registry.go

```diff
@@ -4,42 +4,37 @@ import (
 	"context"
 	"encoding/base64"
 	"fmt"
-	"io"
 	"io/ioutil"
-	"os"
+	"log"
+	"net"
+	"net/http"
+	"net/http/httptest"
 	"path/filepath"
+	"strconv"
 	"testing"
-	"time"
 
-	"github.com/docker/docker/api/types"
-	"github.com/docker/docker/api/types/container"
-	"github.com/docker/go-connections/nat"
-	"golang.org/x/crypto/bcrypt"
-
-	volumetypes "github.com/docker/docker/api/types/volume"
-	"github.com/docker/docker/errdefs"
+	"github.com/google/go-containerregistry/pkg/registry"
 )
 
 type DockerRegistry struct {
 	Host            string
 	Port            string
 	Name            string
+	server          *httptest.Server
 	DockerDirectory string
 	username        string
 	password        string
-	volumeName      string
+	regHandler      http.Handler
+	authnHandler    http.Handler
 }
 
-var registryImageName = "micahyoung/registry:latest"
-
 type RegistryOption func(registry *DockerRegistry)
 
-//WithSharedStorageVolume allows two instances to share the same data volume.
+//WithSharedHandler allows two instances to share the same data by re-using the registry handler.
 //Use an authenticated registry to write to a read-only unauthenticated registry.
-//Volumes that don't exist will be created, then removed on Stop().
-func WithSharedStorageVolume(volumeName string) RegistryOption {
+func WithSharedHandler(handler http.Handler) RegistryOption {
```

Comment:
While we're here I wonder if it would make sense to rename `WithAuth` to `WithDockerConfig`. For some reason I find this easier to understand but maybe it's just me!

Reply:
Actually I realized that then this would read `ih.NewDockerRegistry(ih.WithDockerConfig(dockerConfigDir), ih.WithSharedHandler(sharedRegHandler))` and the extra docker might be redundant. I don't know... I'm just speaking aloud here as I'm trying to remember how this code works.

Reply:
I feel like `WithDockerConfig` might miss that the new randomly generated username and password are also created (which later are used to branch on whether to be read-only or not). Perhaps `WithAuthAndConfig` or something?